### PR TITLE
Hiding

### PR DESCRIPTION
**Pull request by Fris0uman**
#### Summary
```SUMMARY: Features "Adds action to hide in appropritate furniture/terrain"```

#### Purpose of change

Flesh out the hiding mechanic introduced with HIDE_PLACE flag

#### Describe the solution

Hiding is an action bound to <kbd>j</kbd> by default.
Stand next to elligible tile, press <kbd>j</kbd> and choose the direction : if the place is elligible for hiding in it, the game checks your size, and if you're small enough you move to the tile and are hidden in it.

|  Furniture   |  Max Size Allowed   | NO_SIGHT | CONNECT |
| :--------: | :--------: | :--------: | :--------: |
| Pillow Fort | Medium | YES | NO
| Brick Kiln | Tiny | NO | NO
| Metal Kiln Empty | Tiny | NO | NO
| Rock Kiln | Tiny | NO | NO
| Rock Forge | Tiny | NO | NO
| Forge | Tiny | NO | NO
| Open Crate | Medium | YES | NO
| Dumpster | Medium | YES | YES
| Dryer | Small | YES | NO
| Washing Machine | Small | YES | NO
| Locker | Small | YES | NO
| Dresser | Medium | YES | NO
| Fridge | Small | YES | NO
| Mail Box | Tiny | YES | NO
| Recycle Bin | Tiny | YES | NO
| Trash Can | Tiny | YES | NO
| Cupboard | Small | YES | NO
| Wardrobe | Medium| YES | NO

I also edited how get_size() works for player, so now size mutation actually changes your size. Instead of always being a medium size creature, players can actually be anything between ~~tiny~~ small and huge.
 - SIZE EFECTS : 
    - This won't affect traps since only tiny creature are immune to some traps
    -  This will affect visibility when crouched : this is probably intended behaviour as the code for crouch check for size of the player and not for the mutations.
    - This might affect probability to get shot, and the proba to hit NPC of non medium size
    - Same for melee combat small player should be harder to hit
    -  This will affect collision with vehicle : a huge player/NPC will have more stoping power than a medium one.

Dumpster are no longer passable, they were briefly passable when I made HIDE_PLACE flag to allow things to actually hide in it, but now it's not necessary anymore so I reverted it back to what it was before.

Walking on a tile with HIDE_PLACE flag doesn't do anything.
Hiding in a tile might be possible even if the tile is impassable, ~~meaning that you can use it to pass through a locker for example.~~ But you can't use it to pass through impassable tile, as you'll be sent back to your entry point when leaving the hiding place.
You can't hide from one impassable hiding place into another impassable hiding place unless it's of the same furniture type and has the CONNECT flag.

#### Additional context

What's left to do : 
- [x] Make sure move cost is correctly applied
- [x] Apllying NO_SIGHT correctly
- [x] Validate the list of HIDE_PLACE furniture
- [x] Info about hiding properties are displayed in extended description
- [x] NPCs can hide in passable tiles
- [x] Monsters have 1/10 chance to hide in any adjacent hide_place tile on their way

The list of furniture/terrain that can hide a creature needs to be validated so feel free to express yourself on what should be added or removed from the list and why. 
Also we need to agree on what are the size constraints and if there should be a NO_SIGHT penalty or not for each one.

The fact that hiding is now an action, and don't just work by walking on a tile, means that it's more complicated to do for AIs. My aim here is to make it possible for an AI to "randomly" get inside a furniture to hide.
Doing the right thing and implementing correct AI decision making on when and where to hide is beyond my abbilities. But the current implementation should allow for someone else to do it : hide is a Creature classe function so it can be applied to player, monster and NPCs.

I'm oppen to suggestion for the default key, currently <kbd>j</kbd>  .

### CODE REVIEW

**Comment by I-am-Erk** [comment]
Is it possible to make it so that when you unhide, you automatically move to the tile you were in before hiding? This would prevent phasing through impassible terrain by hiding in it.

**Comment by Fris0uman** [comment]
I'm not sure how to store your previous position until it's usefull, but that sounds like a pretty good idea.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/so-you-want-to-hide-in-a-mail-box/19144/1

**Comment by dissociativity** [comment]
This seems pretty cool, but I'm not sure how even a tiny mouse player could hide in a mail-box since most have roughly the volume of an arm, and I'd make locker and fridge small. lockers definitely aren't small enough for anyone but the size of a child to fit in, same with fridges.

**Comment by Fris0uman** [comment]
Fridge is already small, I made a typo in the table. Small for locker seems reasonable.
Yeah mail box is a bit on the comic sides of things, but tiny is suppose to be the size of a squirrel, i'm pretty sure a squirrel can fit in a mail box.

**Comment by kevingranade** [comment]
No way is a player hitting squirrel sized, if there are descriptions
indicating that, they need to be fixed.
Smallest player size would probably be toddler sized or so, human
proportions but about 2-3 ft tall.

**Comment by Fris0uman** [comment]
So as far as MS_SIZE is concerned the lowest size available to player is MS_SMALL and never MS_TINY ?
I'll just edit get_size() to reflect that.

**Comment by Fris0uman** [comment]
So right now zombie just disapear from the map if they try to hide, do monster despawn if they are on an impassable tile ?

**Comment by kevingranade** [comment]
That sounds about right, the game engine strongly assumes that creatures do not exist on impassible tiles.

**Comment by Fris0uman** [comment]
So after more tests it turns out that the game does not despawn monster on impassable tile but teleport them away because they are stuck. 
It might be actually very important to implement I-am-Erk's idea of returning the creature to it's previous position when unhiding because otherwise the AI will just use hide to pass through furnitures and althought because it might solve the teleporting issue. 
I looked at how peek() does it but I don't think it helps much because hide() is just called once to move you so if it stores your previous position the same way peek() does the info is going to be lost when we get out of the hide() loop .
So what would be an acceptable way to store your position when you use hide() the first time and then use it when you try to move out your hiding place ? (And then replacing it whith a new value next time you hide() )

Also I noticed that #28761 makes a nice use of the hiding property of improvised shelter and this PR might break that if AI ability to hide is not correctly implemented.

**Comment by kevingranade** [comment]
Having terrain or furniture marked impassible shouldn't be something you can override.  Is it really necessary for the hiding feature to work?

If you think small creatures should be able to enter squares that are currently impassable, that's a different feature, I'd prefer the hiding thing to just remain movement based.

**Comment by Fris0uman** [comment]
It's not neccessary in fact  the hiding feature is already in place as pillow fort, dumpster and improvised shelter will hide any creature stepping on them.
I just thougth that some furniture like wardrobe and locker could be hiding places too but they are impassable and so the way I found to have them be impassable and hiding places at the same time was to make that hide function that moves you on the tile if you have the right size.  
But if that's not acceptable I can scrap that idea and take time to think some more.

**Comment by kevingranade** [comment]
If nothing else I'm pretty sure that it's going to be difficult to make the code happy with creatures occupying impassible tiles, maybe instead you can remove impassible from these articles and add a new flag that's "impassable_unless_small" or something, though I think that's also going to be pretty tricky.

**Comment by Fris0uman** [comment]
So now monster can hide in impassable tiles without being teleported away, and when  players step out of hiding they are sent back to the position from which they entered the hiding place.

**Comment by Fris0uman** [comment]
So what do we have at this point : 

- NPC will hide in hide-place tile they spawn on so it won't break map build with the previous system
- NPC can hide in passable hide_place : 50% chance to walk on it and 50% to hide in it

- Monster have 10% chance to hide in any size-appropriate hide_place tile and 90% to bash/walk on it

- When hidden in a passable hide_place tile any creature can just walk out of it in any direction at a move cost of 50 on top of the cost of moving

- When hidden in an impassable tile any creature will be sent sent back to its entry point when trying to leave the hide_place. 
Meaning that if you hide() from one open crate to another 10 times when you decide to get out you'll be sent to the tile you entered the chain of crate from.
But if you do the same in a giant pillow fort you will just step out.

- If the tile a creature is hidden in is destroyed the hidden and no_sight effect are removed.

I'm not going to make the NPC hide in impassable tile because it's pointless since all I can do is make the AI hide in a tile on its way to somewhere else and hiding in impassable tile don't get you closer to your goal.
I did it for monsters because it came with the ability to hide in passable tiles, but it's a bit more complicated for NPCs.

So if the list of hide_place funiture and their associated properties are fine with everyone I think I m going to wrap this up.

File: src/creature.cpp [hunk 1]
```diff
@@ -168,16 +168,22 @@ bool Creature::hide( const tripoint &target, bool move )
 
 bool Creature::unhide( const tripoint entered_from )
 {
-    if( has_effect( effect_no_sight ) && !g->m.has_flag_ter_or_furn( TFLAG_NO_SIGHT, entered_from ) ) {
+    if( has_effect( effect_no_sight ) ) {
```
Comment (Night-Pryanik) [inline src/creature.cpp]: With no other conditions this check is redundant. The same with `effect_hidden` below.

Comment (Fris0uman) [reply src/creature.cpp]: Done

**Comment by Fris0uman** [comment]
I think it's done at this point. 
It would be cool to add "Press <key bound to Hide> to hide" to the extended description but I don't know how to do it.

**Comment by Yunatan** [comment]
> Locker | Small

I think lockers should allow medium size just as a "rule of cool" because they are the most abundant furniture on the list and it is actually justifiable.

> I also edited how get_size() works for player, so now size mutation actually changes your size. Instead of always being a medium size creature, players can actually be anything between tiny small and huge.

Have you considered all the balance implications of that? From the top of my head I can name the [carrying capacity adjustments](http://cddawiki.chezzo.com/cdda_wiki/index.php?title=Weight#Carrying_Capacity) and [some traps interactions](http://cddawiki.chezzo.com/cdda_wiki/index.php?title=Traps#List_of_traps). It would be wise to include everything that is affected by this change into the changelog just as a heads-up for the players. 

> Hiding in a tile might be possible even if the tile is impassable. But you can't use it to pass through impassable tile, as you'll be sent back to your entry point when living the hiding place.

As you [said in a different issue](https://github.com/CleverRaven/Cataclysm-DDA/issues/29891#issuecomment-486289543), this allows to hop from locker to locker and then be sent back to the point you entered the lockers from. So there probably should be a distinction when you can re-hide into the adjacent appropriate furniture/terrain (e.g. interconnected pillow fort or bushes) and when you must leave your current hiding place first (e.g. lockers and crates). 

> I'm open to suggestion for the default key, currently j .

Why not make a part of an `e`xamine action?

**Comment by Fris0uman** [comment]
> rule of cool

It is very rarely (if ever) a rule used to make decision regarding balance of the game.

> mutation actually afecting size

You're right I need to look a bit more into what's happening with that. But I think it makes sense that size mutations actually affect size.

> hide hopping

I just made some change to that : now you can't hide into an adjacent hide_place if it's impassable unless it is a tile of the same type AND has the CONNECT flag.
Basically you can't hop from impassable hide_place to another except between dumpsters because i think they are suppose to represent a bigger multi tile dumpster.

> `e`xamine
I'll look into doing that

File: doc/JSON_FLAGS.md [hunk 2]
```diff
@@ -63,6 +63,7 @@ List of known flags, used in both terrain.json and furniture.json
 - ```CAN_SIT``` Furniture the player can sit on. Player sitting near furniture with the "FLAT_SURF" tag will get mood bonus for eating.
 - ```CHIP``` Used in construction menu to determine if wall can have paint chipped off.
 - ```COLLAPSES``` Has a roof that can collapse.
+- ```CONNECT```` Only for impassable tile with the HIDE_PLACE flag, allows to hide into this tile from a similar one. Does not affect the way this tile is displayed !
```
Comment (Night-Pryanik) [inline doc/JSON_FLAGS.md]:
```suggestion
- ```CONNECT```` Only for impassable tile with the HIDE_PLACE flag, allows to hide into this tile from a similar one. Does not affect the way this tile is displayed!
```

File: src/creature.cpp [hunk 3]
```diff
@@ -139,6 +140,75 @@ void Creature::process_turn()
...
+            return false;
+        }
+    } else if( g->m.has_flag_ter_or_furn( "MAX_LARGE", target ) ) {
+        if( get_size() > MS_LARGE ) {
+            return false;
+        }
+    }
+
+    if( has_effect( effect_hidden ) && g->m.impassable( target ) ) {
+        if( ( g->m.ter( pos() ) == g->m.ter( target ) ) ) {
```
Comment (Night-Pryanik) [inline src/creature.cpp]:
```suggestion
        if( g->m.ter( pos() ) == g->m.ter( target ) ) {
```

File: src/creature.cpp [hunk 4]
```diff
@@ -139,6 +140,75 @@ void Creature::process_turn()
...
+            if( !g->m.has_flag_ter_or_furn( "CONNECT", target ) ) {
+                return false;
+            }
+        } else {
+            return false;
+        }
+    }
+
```
Comment (Night-Pryanik) [inline src/creature.cpp]:
```suggestion
```

File: src/creature.cpp [hunk 5]
```diff
@@ -139,6 +140,75 @@ void Creature::process_turn()
...
+
+    if( move ) {
+        setpos( target );
+        moves -= 100; // TODO : make cost specific to hiding
```
Comment (Night-Pryanik) [inline src/creature.cpp]:
```suggestion
        mod_moves( -100 ); // TODO : make cost specific to hiding
```

File: src/creature.cpp [hunk 6]
```diff
@@ -139,6 +140,75 @@ void Creature::process_turn()
...
+        add_effect( effect_hidden, 1_turns, num_bp, true );
+        if( is_player() ) {
+            add_msg( m_good, _( "You are hiding in the %s." ), g->m.name( target ).c_str() );
```
Comment (Night-Pryanik) [inline src/creature.cpp]:
```suggestion
            add_msg( m_good, _( "You are hiding in the %s." ), g->m.name( target ) );
```

File: src/creature.cpp [hunk 7]
```diff
@@ -139,6 +140,75 @@ void Creature::process_turn()
...
+        }
+        if( g->m.has_flag_ter_or_furn( TFLAG_NO_SIGHT, target ) ) {
+            add_effect( effect_no_sight, 1_turns, num_bp, true );
+        }
+    }
+
+    return true;
+
```
Comment (Night-Pryanik) [inline src/creature.cpp]:
```suggestion
```

File: src/creature.cpp [hunk 8]
```diff
@@ -139,6 +140,75 @@ void Creature::process_turn()
...
+}
+
+bool Creature::unhide( const tripoint entered_from )
+{
+    remove_effect( effect_no_sight );
+    remove_effect( effect_hidden );
+
+    if( !g->m.impassable( pos() ) ) {
+        moves -= 50; // cost less move because we're not moving, just unhiding.
```
Comment (Night-Pryanik) [inline src/creature.cpp]:
```suggestion
        mod_moves( -50 ); // cost less move because we're not moving, just unhiding.
```

File: src/creature.cpp [hunk 9]
```diff
@@ -139,6 +140,75 @@ void Creature::process_turn()
...
+        return true;
+    }
+
+    setpos( entered_from );
+    moves -= ( g->m.move_cost( entered_from ) +
```
Comment (Night-Pryanik) [inline src/creature.cpp]:
```suggestion
    mod_moves ( -g->m.move_cost( entered_from ) + 100 );
```

File: src/descriptions.cpp [hunk 10]
```diff
@@ -177,5 +177,42 @@ std::string map_data_common_t::extended_description() const
         ss << std::endl;
     }
 
+    if( has_flag( "HIDE_PLACE" ) ) {
+        ss << "--" << std::endl;
+        ss << "This could be use as an hiding place." << std::endl;
+        if( has_flag( "MAX_TINY" ) ) {
+            ss << "This place can accommodate <info>Tiny</info> creatures." << std::endl;
```
Comment (Night-Pryanik) [inline src/descriptions.cpp]:
```suggestion
            ss << "This place can accommodate <info>tiny</info> creatures." << std::endl;
```

File: src/descriptions.cpp [hunk 11]
```diff
@@ -177,5 +177,42 @@ std::string map_data_common_t::extended_description() const
...
+            if( g->u.get_size() == MS_TINY ) {
+                ss << "<good>Your size allows you to hide here.</good>" << std::endl;
+            } else {
+                ss << "<bad>You are too big to hide in here.</bad>" << std::endl;
+            }
+        } else if( has_flag( "MAX_SMALL" ) ) {
+            ss << "This place can accommodate <info>Small</info> creatures." << std::endl;
```
Comment (Night-Pryanik) [inline src/descriptions.cpp]:
```suggestion
            ss << "This place can accommodate <info>small</info> creatures." << std::endl;
```

File: src/descriptions.cpp [hunk 12]
```diff
@@ -177,5 +177,42 @@ std::string map_data_common_t::extended_description() const
...
+            if( g->u.get_size() <= MS_SMALL ) {
+                ss << "<good>Your size allows you to hide here.</good>" << std::endl;
+            } else {
+                ss << "<bad>You are too big to hide in here.</bad>" << std::endl;
+            }
+        } else if( has_flag( "MAX_MEDIUM" ) ) {
+            ss << "This place can accommodate <info>Medium</info> creatures." << std::endl;
```
Comment (Night-Pryanik) [inline src/descriptions.cpp]:
```suggestion
            ss << "This place can accommodate <info>medium</info> creatures." << std::endl;
```

File: src/descriptions.cpp [hunk 13]
```diff
@@ -177,5 +177,42 @@ std::string map_data_common_t::extended_description() const
...
+            if( g->u.get_size() <= MS_MEDIUM ) {
+                ss << "<good>Your size allows you to hide here.</good>" << std::endl;
+            } else {
+                ss << "<bad>You are too big to hide in here.</bad>" << std::endl;
+            }
+        } else if( has_flag( "MAX_LARGE" ) ) {
+            ss << "This place can accommodate <info>Large</info> creatures." << std::endl;
```
Comment (Night-Pryanik) [inline src/descriptions.cpp]:
```suggestion
            ss << "This place can accommodate <info>large</info> creatures." << std::endl;
```

File: src/descriptions.cpp [hunk 14]
```diff
@@ -177,5 +177,42 @@ std::string map_data_common_t::extended_description() const
...
+            if( g->u.get_size() <= MS_LARGE ) {
+                ss << "<good>Your size allows you to hide here.</good>" << std::endl;
+            } else {
+                ss << "<bad>You are too big to hide in here.</bad>" << std::endl;
+            }
+
```
Comment (Night-Pryanik) [inline src/descriptions.cpp]:
```suggestion
```

File: src/handle_action.cpp [hunk 15]
```diff
@@ -1598,6 +1599,23 @@ bool game::handle_action()
                 }
                 break;
 
+            case ACTION_HIDE:
+                if( const cata::optional<tripoint> pnt = choose_adjacent( _( "Hide where?" ) ) ) {
+                    if( !u.hide( *pnt, false ) ) {
+                        add_msg( m_info, _( "You can't hide in the %s." ), g->m.name( *pnt ).c_str() );
```
Comment (Night-Pryanik) [inline src/handle_action.cpp]:
```suggestion
                        add_msg( m_info, _( "You can't hide in the %s." ), g->m.name( *pnt ) );
```

File: src/npcmove.cpp [hunk 16]
```diff
@@ -1805,8 +1806,34 @@ void npc::move_to( const tripoint &pt, bool no_bashing, std::set<tripoint> *nomo
 
         moves = 0;
     }
+    // If we can hide in it, we might do so
+    // TODO : Make hiding more sensible and less random
+    if( hide( p, false ) ) {
+        if( one_in( 2 ) ) {
+
```
Comment (Night-Pryanik) [inline src/npcmove.cpp]:
```suggestion
```

File: src/npcmove.cpp [hunk 17]
```diff
@@ -1805,8 +1806,34 @@ void npc::move_to( const tripoint &pt, bool no_bashing, std::set<tripoint> *nomo
...
+            if( !has_effect( effect_hidden ) ) {
+                prev_pos = pos();
+            }
+            hide( p );
+            return;
+        }
+    }
 
     if( moved ) {
+
```
Comment (Night-Pryanik) [inline src/npcmove.cpp]:
```suggestion
```

**Comment by Fris0uman** [comment]
So after messing around a bit with iexamine, I'm not sure it's a good idea to add hiding as an `e`xamine action because it might become very anoying to go through the menu each time you try to get something from a wardrob or a locker.

**Comment by Fris0uman** [comment]
The crouching system already does most of the hiding stuff, so I'm just going to re-submit the part about player size and let the hiding in furniture idea rest for now.
